Dockerfile

```diff
@@ -1,95 +1,91 @@
+# Build VueJS front-end
+RUN (cd plugins/magma; npm install && npm run build)
+
+# This is the runtime stage
+# It containes all dependencies required by caldera
+FROM debian:bookworm-slim as runtime
+
+# There are two variants - slim and full
+# The slim variant excludes some dependencies of *emu* and *slim* that can be downloaded on-demand if needed
+# They are very large
+ARG VARIANT=full
+ENV VARIANT=${VARIANT}
+
+# Display an error if variant is set incorrectly, otherwise just print information regarding which variant is in use
+RUN if [ "$VARIANT" = "full" ]; then \
+        echo "Building \"full\" container suitable for offline use!"; \
+    elif [ "$VARIANT" = "slim" ]; then \
+        echo "Building slim container - some plugins (emu, atomic) may not be available without an internet connection!"; \
+    else \
+        echo "Invalid Docker build-arg!"; \
+        exit 1; \
+fi
 
-# Compile default sandcat agent binaries, which will download basic golang dependencies.
+WORKDIR /usr/src/app
 
-# Install Go dependencies
-WORKDIR /usr/src/app/plugins/sandcat/gocat
-RUN go mod tidy && go mod download
+# Copy in source code and compiled UI
+# IMPORTANT NOTE: the .dockerignore file is very important in preventing weird issues.
+# Especially if caldera was ever compiled outside of Docker - we don't want those files to interfere with this build process,
+# which should be repeatable.
+ADD . .
+COPY --from=ui-build /usr/src/app/plugins/magma/dist /usr/src/app/plugins/magma/dist
 
-WORKDIR /usr/src/app/plugins/sandcat
+# From https://docs.docker.com/build/building/best-practices/
+# Install caldera dependencies
+RUN apt-get update && \
+apt-get --no-install-recommends -y install git curl unzip python3-dev python3-pip golang-go mingw-w64 zlib1g gcc && \
+rm -rf /var/lib/apt/lists/*
 
 # Fix line ending error that can be caused by cloning the project in a Windows environment
-RUN if [ "$WIN_BUILD" = "true" ] ; then cp ./update-agents.sh ./update-agents-copy.sh; fi
-RUN if [ "$WIN_BUILD" = "true" ] ; then tr -d '\15\32' < ./update-agents-copy.sh > ./update-agents.sh; fi
-RUN if [ "$WIN_BUILD" = "true" ] ; then rm ./update-agents-copy.sh; fi
-
-RUN ./update-agents.sh
-
-# Check if we can compile the sandcat extensions, which will download golang dependencies for agent extensions
-RUN mkdir /tmp/gocatextensionstest
-
-RUN cp -R ./gocat /tmp/gocatextensionstest/gocat
-RUN cp -R ./gocat-extensions/* /tmp/gocatextensionstest/gocat/
+RUN cd /usr/src/app/plugins/sandcat; tr -d '\15\32' < ./update-agents.sh > ./update-agents.sh
 
-RUN cp ./update-agents.sh /tmp/gocatextensionstest/update-agents.sh
-
-WORKDIR /tmp/gocatextensionstest
-
-RUN mkdir /tmp/gocatextensionstest/payloads
+# Set timezone (default to UTC)
+ARG TZ="UTC"
+RUN ln -snf /usr/share/zoneinfo/$TZ /etc/localtime && \
+    echo $TZ > /etc/timezone
 
-RUN ./update-agents.sh
+# Install pip requirements
+RUN pip3 install --break-system-packages --no-cache-dir -r requirements.txt 
```

Comment:
thoughts on installing packages within a virtual env?

Reply:
This was a conscious choice. We already are in a Docker container, we don't have to worry about polluting anything.